### PR TITLE
Add single asset exits for Weighted pools in generalised flow

### PR DESCRIPTION
**Pull request by garethfuller**
# Description

**When exiting deep Weighted pool, limit single asset tokens to high level tokens.**

For details on the functions that are be used for exits see:
- https://github.com/balancer-labs/balancer-sdk/blob/develop/balancer-js/README.md#buildexitexactbptin
- https://github.com/balancer-labs/balancer-sdk/blob/develop/balancer-js/README.md#buildexitexacttokensout
And for price impact:
- https://github.com/balancer-labs/balancer-sdk/blob/develop/balancer-js/examples/priceImpact.ts

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?

- Go to Deep Weighted Pool's withdraw page `http://localhost:8080/#/ethereum/pool/0x25accb7943fd73dda5e23ba6329085a3c24bfb6a000200000000000000000387/withdraw`
- Select "Single token" tab
- Select any pool token to exit to and try inputting different amounts

## Visual context
Screenshot 2023-01-04 at 12 25 17

## Visual context

Please provide any relevant visual context for UI changes or additions. This could be static screenshots or a loom screencast.

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have requested at least 1 review (If the PR is significant enough, use best judgement here)
- [ ] I have commented my code where relevant, particularly in hard-to-understand areas
- [ ] If package-lock.json has changes, it was intentional.
- [ ] The base of this PR is `master` if hotfix, `develop` if not

### CODE REVIEW

File: src/providers/local/exit-pool.provider.ts [hunk 1]
```diff
@@ -172,6 +185,21 @@ const provider = (props: Props) => {
...
+    (): boolean =>
+      isSingleAssetExit.value &&
+      isDeep(pool.value) &&
+      isPreMintedBptType(pool.value.poolType)
+  );
+
+  const exitHandlerType = computed((): ExitHandler => {
+    if (shouldUseSwapExit.value) return ExitHandler.Swap;
+    if (isWeightedPool.value && isSingleAssetExit.value)
+      return ExitHandler.ExactOut;
```
Comment (garethfuller) [inline src/providers/local/exit-pool.provider.ts:198]: Why did you change this logic? If the user has clicked the max button we need to do ExactIn so that the full BPT balance is used. Otherwise, you can end up with a dust situation.

File: src/services/balancer/pools/exits/exit-pool.service.ts [hunk 2]
```diff
@@ -33,28 +47,32 @@ export class ExitPoolService {
...
+      case ExitHandler.Generalised:
+        return (this.exitHandler = new GeneralisedExitHandler(
+          ...handlerParams
+        ));
+      case ExitHandler.ExactIn:
+        return (this.exitHandler = new ExactInExitHandler(...handlerParams));
+      case ExitHandler.ExactOut:
+        return (this.exitHandler = new ExactOutExitHandler(...handlerParams));
+      default:
+        throw new Error(`Pool type not handled: ${pool.value.poolType}`);
```
Comment (garethfuller) [inline src/services/balancer/pools/exits/exit-pool.service.ts:75]:
```suggestion
        throw new Error(`Exit type not handled: ${type}`);
```

**Review by garethfuller — COMMENTED** [review]
@simeonkerkola First pass, still need to test UI.

File: src/providers/local/exit-pool.provider.ts (hunk 1 above)
Comment (simeonkerkola) [reply src/providers/local/exit-pool.provider.ts:199]: Right 👍. Changed it back.

**Comment by timjrobinson** [comment]
Tested using mainnet and this worked for me. Was able to deposit wstETH into https://add-exit-handlers.balancer-d6h.pages.dev/#/ethereum/pool/0x25accb7943fd73dda5e23ba6329085a3c24bfb6a000200000000000000000387 and withdraw as bb-a-USD and it worked.

File: src/services/balancer/pools/exits/handlers/exact-in-exit.handler.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      tokenOut
+    );
+
+    return {
+      amountsOut: { [tokenOutAddress]: scaledAmountOut },
+      priceImpact: Number(scaledPriceImpact),
+    };
+  }
+
+  private getScaledAmountOut(
```
Comment (garethfuller) [inline src/services/balancer/pools/exits/handlers/exact-in-exit.handler.ts:94]: The naming here isn't is clear as it could be. Maybe `normalizeAmountOut` would be better since you're converting a evm (scaled) amount to a normal amount.

File: src/services/balancer/pools/exits/handlers/exact-in-exit.handler.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    // Because this is an exit we need to pass amountsOut as the amountsIn and
+    // bptIn as the minBptOut to this calcPriceImpact function.
+    const priceImpact = await sdkPool.calcPriceImpact(
+      expectedAmountsOut,
+      evmBptIn,
+      false
+    );
+
+    const scaledPriceImpact = formatFixed(priceImpact, 18);
+    const scaledAmountOut = this.getScaledAmountOut(
```
Comment (garethfuller) [inline src/services/balancer/pools/exits/handlers/exact-in-exit.handler.ts:82]: Then this variable would become:
```suggestion
    const normalizedAmountOut = this.normalizeAmountOut(
```

File: src/services/balancer/pools/exits/handlers/exact-in-exit.handler.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Because this is an exit we need to pass amountsOut as the amountsIn and
+    // bptIn as the minBptOut to this calcPriceImpact function.
+    const priceImpact = await sdkPool.calcPriceImpact(
+      expectedAmountsOut,
+      evmBptIn,
+      false
+    );
+
+    const scaledPriceImpact = formatFixed(priceImpact, 18);
```
Comment (garethfuller) [inline src/services/balancer/pools/exits/handlers/exact-in-exit.handler.ts:81]: Also, I'd maybe change this to:
```suggestion
    const evmPriceImpact = await sdkPool.calcPriceImpact(
      expectedAmountsOut,
      evmBptIn,
      false
    );

    const priceImpact = formatFixed(evmPriceImpact, 18);
```

File: src/services/balancer/pools/exits/handlers/exact-out-exit.handler.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    // Because this is an exit we need to pass amountsOut as the amountsIn and
+    // bptIn as the minBptOut to this calcPriceImpact function.
+    const priceImpact = await sdkPool.calcPriceImpact(
+      fullAmountsOut,
+      this.lastExitRes.expectedBPTIn,
+      false
+    );
+
+    const scaledPriceImpact = formatFixed(priceImpact, 18);
```
Comment (garethfuller) [inline src/services/balancer/pools/exits/handlers/exact-out-exit.handler.ts:78]: Same here:
```suggestion
    const evmPriceImpact = await sdkPool.calcPriceImpact(
      fullAmountsOut,
      this.lastExitRes.expectedBPTIn,
      false
    );

    const priceImpact = formatFixed(evmPriceImpact, 18);
```

**Review by garethfuller — COMMENTED** [review]
@simeonkerkola
